### ISSUE TITLE
UPSERT support?

### ISSUE THREAD

**Issue by lastmeta** [serverpod/serverpod#232]
Postgresql has a cool ability to insert, but if there's a conflict then just update instead, known as Upsert.

https://wiki.postgresql.org/wiki/UPSERT
https://stackoverflow.com/questions/17267417/how-to-upsert-merge-insert-on-duplicate-update-in-postgresql

I see a function for `.insert` and `.update` but no `.upsert` and I went and looked at insert and it doesn't work like upsert by default.

I think it'd be a cool feature to have `.upsert` so I don't have to query first to see if it exists, then query again to put it in.

It would be cool but for now I'll do it the long way.

**Comment by vlidholt** [comment]
Will add this feature!

**Comment by II11II** [comment]
Should the serverpod's orm update on conflict?

**Comment by lukehutch** [comment]
So I'm assuming the `id` field will be ignored by an `upsert` operation, even if `id` is set, given that Postgres doesn`t actually have an `UPSERT` command, but rather uses `INSERT ... ON CONFLICT ... UPDATE` command?

**Comment by SandPod** [comment]
Draft implementation and source of inspiration: https://github.com/serverpod/serverpod/pull/580.

If above PR is used as inspiration for the implementing this feature, the author @sedobrengocce should be credited as a co-author.

**Comment by t-kietzmann** [comment]
Is this feature being actively worked on?

**Comment by iampopal** [comment]
We need this feature too.

**Comment by lastmeta** [comment]
> @vlidholt @lastmeta Should the serverpod's orm update on conflict?

I have implemented on conflict before which I would assume is what upsert is a shorthand for. just randomly guessing here, but I think the api would look like .upsert(<options to modify how ON CONFLICT should work>) and it should probably use on conflict in the actual sql. but that's totally a guess. idk.

**Comment by lukehutch** [comment]
Correct, in Postgres you insert, then on conflict, update instead. This is called upset in other databases.

**Comment by iampopal** [comment]
Yes other orm also do this. They try insert … on conflict..

**Comment by iampopal** [comment]
Updates...

**Comment by SandPod** [comment]
We are not actively working on this feature and have no updates to share at this moment.

**Comment by parnesen** [comment]
me need too. I have a table where each row might contain two largish blobs. I really don't want to have to load it into memory just to see if it exists.

**Comment by christerswahn** [comment]
An additional benefit of a good `upsert` feature will be improved robustness within transactions: If an INSERT within a transaction fails, the transaction is _aborted_ in PostreSQL regardless of whether the code handles the resulting exception in an intentional manner or not, and has to be rolled back.

But if the INSERT has an ON CONFLICT clause, the query will not raise a conflict error. This could be used as either ON CONFLICT DO NOTHING or ON CONFLICT UPDATE, and the transaction may continue.

The ORM insert or upsert method would need to report if an actual insert or update (or no action) took place, or make it deducable from the returned object's content.

**Comment by v1ktorstep** [comment]
It's sad that we don't have this feature yet.

**Comment by sedobrengocce** [comment]
I'm sorry for not completing this feature during the last 3 yr, now i created a pull request  (#4832 )that pass all tests and works. I hope it will be merged soon

**Comment by marcelomendoncasoares** [comment]
> I'm sorry for not completing this feature during the last 3 yr, now i created a pull request that pass all tests and works. I hope it will be merged soon

We'll work to have this in very soon!
